### ISSUE TITLE
[Bug]: Can't compile go programs in debug mode

### ISSUE THREAD

**Issue by ibuildthecloud** [jetpack-io/devbox#1215]
**Current Behavior (bug)**
Ironically I'm trying to hack on devbox inside of devbox and I hit this issue.

```
$ ~/go/bin/dlv debug ./cmd/devbox/
# runtime/cgo
In file included from /nix/store/bhhb91angi23wsr42apz1in1n9b9ilg3-glibc-2.37-8-dev/include/bits/libc-header-start.h:33,
                 from /nix/store/bhhb91angi23wsr42apz1in1n9b9ilg3-glibc-2.37-8-dev/include/stdlib.h:26,
                 from _cgo_export.c:3:
/nix/store/bhhb91angi23wsr42apz1in1n9b9ilg3-glibc-2.37-8-dev/include/features.h:413:4: error: #warning _FORTIFY_SOURCE requires compiling with optimization (-O) [-Werror=cpp]
  413 | #  warning _FORTIFY_SOURCE requires compiling with optimization (-O)
      |    ^~~~~~~
cc1: all warnings being treated as errors
exit status 1
```

I know this is really a nix issue and not devbox directly, but being that this configuration of running `go` sdk inside devbox is a primary use case I would assume this would work.

**Additional context**
Please include the output of `devbox version -v` and
```
$ devbox version -v
Version:     0.5.5
Platform:    linux_amd64
Commit:      891e8de7a8ab7a3fec42fad069326d68cb84603e
Commit Time: 2023-06-19T19:18:00Z
Go Version:  go1.20.5
Launcher:    0.2.0
```
```
$ cat /home/darren/.local/share/devbox/global/default/devbox.json
{
  "packages": [
    "vim-full@latest",
    "cmake@latest",
    "python@3.8",
    "gcc@latest",
    "libxcrypt@latest",
    "go@1.20",
    "gh@latest",
    "k3d@latest",
    "xdg-utils@latest",
    "xfce.xfce4-terminal@latest",
    "ubuntu_font_family@latest",
    "hicolor-icon-theme@latest",
    "fontconfig@latest",
    "xorg.xorgserver@latest",
    "xterm@latest",
    "jetbrains.goland@latest",
    "fuse3@latest",
    "fuse@latest",
    "strace@latest",
    "git@latest",
    "procps@latest"
  ],
  "shell": {
    "init_hook": [
      "echo 'Welcome to devbox!' > /dev/null"
    ],
    "scripts": {
      "test": [
        "echo \"Error: no test specified\" && exit 1"
      ]
    }
  }
}

$ cat /home/darren/.local/share/devbox/global/default/devbox.lock
{
  "lockfile_version": "1",
  "packages": {
    "cmake@latest": {
      "last_modified": "2023-05-01T16:53:22Z",
      "resolved": "github:NixOS/nixpkgs/8670e496ffd093b60e74e7fa53526aa5920d09eb#cmake",
      "version": "3.25.3"
    },
    "fontconfig@latest": {
      "last_modified": "2023-05-01T16:53:22Z",
      "resolved": "github:NixOS/nixpkgs/8670e496ffd093b60e74e7fa53526aa5920d09eb#fontconfig",
      "version": "2.14.0"
    },
    "fuse3@latest": {
      "last_modified": "2023-05-01T16:53:22Z",
      "resolved": "github:NixOS/nixpkgs/8670e496ffd093b60e74e7fa53526aa5920d09eb#fuse3",
      "version": "3.11.0"
    },
    "fuse@latest": {
      "last_modified": "2023-05-01T16:53:22Z",
      "resolved": "github:NixOS/nixpkgs/8670e496ffd093b60e74e7fa53526aa5920d09eb#fuse",
      "version": "2.9.9"
    },
    "gcc@latest": {
      "last_modified": "2023-05-01T16:53:22Z",
      "resolved": "github:NixOS/nixpkgs/8670e496ffd093b60e74e7fa53526aa5920d09eb#gcc",
      "version": "12.2.0"
    },
    "gh@latest": {
      "last_modified": "2023-05-01T16:53:22Z",
      "resolved": "github:NixOS/nixpkgs/8670e496ffd093b60e74e7fa53526aa5920d09eb#gh",
      "version": "2.28.0"
    },
    "git@latest": {
      "last_modified": "2023-05-01T16:53:22Z",
      "resolved": "github:NixOS/nixpkgs/8670e496ffd093b60e74e7fa53526aa5920d09eb#git",
      "version": "2.40.1"
    },
    "go@1.20": {
      "last_modified": "2023-05-01T16:53:22Z",
      "resolved": "github:NixOS/nixpkgs/8670e496ffd093b60e74e7fa53526aa5920d09eb#go",
      "version": "1.20.3"
    },
    "hicolor-icon-theme@latest": {
      "last_modified": "2023-05-01T16:53:22Z",
      "resolved": "github:NixOS/nixpkgs/8670e496ffd093b60e74e7fa53526aa5920d09eb#hicolor-icon-theme",
      "version": "0.17"
    },
    "jetbrains.goland@latest": {
      "last_modified": "2023-05-01T16:53:22Z",
      "resolved": "github:NixOS/nixpkgs/8670e496ffd093b60e74e7fa53526aa5920d09eb#jetbrains.goland",
      "version": "2023.1"
    },
    "k3d@latest": {
      "last_modified": "2022-01-01T22:37:24Z",
      "resolved": "github:NixOS/nixpkgs/0fd74961da9fe2640ad2fbd960c3af08246e6749#k3d",
      "version": "0.8.0.6"
    },
    "libxcrypt@latest": {
      "last_modified": "2023-05-01T16:53:22Z",
      "resolved": "github:NixOS/nixpkgs/8670e496ffd093b60e74e7fa53526aa5920d09eb#libxcrypt",
      "version": "4.4.33"
    },
    "procps@latest": {
      "last_modified": "2023-05-01T16:53:22Z",
      "resolved": "github:NixOS/nixpkgs/8670e496ffd093b60e74e7fa53526aa5920d09eb#procps",
      "version": "3.3.17"
    },
    "python@3.8": {
      "last_modified": "2023-05-01T16:53:22Z",
      "plugin_version": "0.0.1",
      "resolved": "github:NixOS/nixpkgs/8670e496ffd093b60e74e7fa53526aa5920d09eb#python38",
      "version": "3.8.16"
    },
    "strace@latest": {
      "last_modified": "2023-05-01T16:53:22Z",
      "resolved": "github:NixOS/nixpkgs/8670e496ffd093b60e74e7fa53526aa5920d09eb#strace",
      "version": "6.2"
    },
    "ubuntu_font_family@latest": {
      "last_modified": "2023-05-01T16:53:22Z",
      "resolved": "github:NixOS/nixpkgs/8670e496ffd093b60e74e7fa53526aa5920d09eb#ubuntu_font_family",
      "version": "0.83"
    },
    "vim-full@latest": {
      "last_modified": "2023-05-01T16:53:22Z",
      "resolved": "github:NixOS/nixpkgs/8670e496ffd093b60e74e7fa53526aa5920d09eb#vim-full",
      "version": "9.0.1441"
    },
    "xdg-utils@latest": {
      "last_modified": "2023-05-01T16:53:22Z",
      "resolved": "github:NixOS/nixpkgs/8670e496ffd093b60e74e7fa53526aa5920d09eb#xdg-utils",
      "version": "2020-10-21"
    },
    "xfce.xfce4-terminal@latest": {
      "last_modified": "2023-05-01T16:53:22Z",
      "resolved": "github:NixOS/nixpkgs/8670e496ffd093b60e74e7fa53526aa5920d09eb#xfce.xfce4-terminal",
      "version": "1.0.4"
    },
    "xorg.xorgserver@latest": {
      "last_modified": "2023-05-01T16:53:22Z",
      "resolved": "github:NixOS/nixpkgs/8670e496ffd093b60e74e7fa53526aa5920d09eb#xorg.xorgserver",
      "version": "21.1.8"
    },
    "xterm@latest": {
      "last_modified": "2023-05-01T16:53:22Z",
      "resolved": "github:NixOS/nixpkgs/8670e496ffd093b60e74e7fa53526aa5920d09eb#xterm",
      "version": "379"
    }
  }
}
```

**Comment by ibuildthecloud** [comment]
This seems to be the fix but I have no clue what this means as I don't know the nix-fu.  From https://nixos.wiki/wiki/Go


If you encounter [this issue](https://github.com/go-delve/delve/issues/3085) and receive an error about _FORTIFY_SOURCE when running delve (for example in VSCode), put hardeningDisable = [ "fortify" ]; inside shell.nix or in the mkShell invocation argument like this:
```
pkgs.mkShell {
  hardeningDisable = [ "fortify" ];
  buildInputs = [ pkgs.go_1_18 ];
};
```

**Comment by jimmidyson** [comment]
This is a real blocker to using devbox for (at least) go development. @ibuildthecloud Did you figure out how to enable this? @savil Any help here? :pray:

**Comment by Lagoja** [comment]
@ibuildthecloud @jimmidyson, I found a way to disable the Nix hardening in your Devbox shell. You need to set the NIX_HARDENING_ENABLE variable to an empty string in the `env` section of your `devbox.json`:

```json
{
  "packages": [
    "go@1.20",
    "golangci-lint@1.52.2",
    "delve@latest"
  ],
  "env": {
    "NIX_HARDENING_ENABLE": ""
  },
  "shell": {
    "init_hook": [
      "export \"GOROOT=$(go env GOROOT)\""
    ],
    "scripts": {...}
  }
}
```

I verified that this let me run `dlv debug ./cmd/devbox/` on WSL2, using Debian. Please try it out and let me know if this helps.

Note that this will disable _all_ of the [hardening features](https://ryantm.github.io/nixpkgs/stdenv/stdenv/#sec-hardening-in-nixpkgs) for the Nix stdenv. If you want to selectively enable some of them, you can add them to the environment variable like: 

```json
  "env": {
    "NIX_HARDENING_ENABLE": "stackprotector format"
  },
```

Adding this to the `devbox.json` is not ideal for production binaries. We're working on a way to set env variables locally on a temporary basis (via a `--env` flag or `.env` file), which can be used for this in the future.

**Comment by jimmidyson** [comment]
@Lagoja Genius! This works perfectly, thank you! As you said, I wouldn't want this for production builds so will have to figure out a nice way to work with `--env` and vs code plugin, perhaps that would be a good change there  to add configuration of env vars just for vs code?

**Comment by savil** [comment]
@mohsenari could the devbox-vscode plugin have a setting to (optionally) set environment variables? In this case, one may want to set `NIX_HARDENING_ENABLE=""` to enable the debugger.

**Comment by mohsenari** [comment]
Yeah it's possible but it requires a vscode window reload (just like open in devbox shell environment command). Now that we have --env PRs out, I can either add a command to reload the environment with taking a .env file or prompt for specific env key-value and reload vscode with that env passed to CLI to include in the new environment before reopening vscode. 
Let's discuss the approach internally. I'll update what we end up doing in this thread.

**Comment by Lagoja** [comment]
The --env feature is now available in 0.5.11. 

@mohsenari I think having a way to take a .env file (maybe parsing it out of the `.envrc`?) would be a pretty nice UX. It might be a good start to focus on loading it from a file first, and then figure out the UX for adding env variables directly on a follow up release

**Comment by Lagoja** [comment]
You can now use the `--env` or `--env-file` paths when generating your .envrc with `devbox generate direnv`. 

The recommendation would be to add a `.env` file with the variables you want to customize locally, and then run `devbox generate direnv --env-file .env` so that direnv picks up those changes automatically when cd'ing into the directory

Please let me know if this does not resolve the issue, or if there's a different direnv workflow that we need to support.
